strax/processing/hitlets.py

```diff
@@ -133,165 +136,90 @@ def refresh_hit_to_hitlets(hits, hitlets):
-    """
-    overlap_hit_i, overlap_record_i = strax.overlap_indices(thing['time']//thing['dt'],
-                                                            thing['length'],
-                                                            container['time']//container['dt'],
-                                                            container['length'])
-    data = container['data'][overlap_record_i[0]:overlap_record_i[1]]
-    return data, overlap_hit_i
+    hitelts_is_single_row = isinstance(hitlets, np.void)
+    if hitelts_is_single_row:
+        # A structured array becomes void type if a single row is called,
+        # e.g. hitlets[0] which does not work in numba while, hitlets[:1]
+        # does. So we have to convert the row into the correct format first.
+        hitlets = np.array([hitlets])
+
+    data_field_in_hitlets = 'data' in hitlets.dtype.names
+    if data_field_in_hitlets:
+        data_is_not_empty = np.any(hitlets['data'] != 0)
+        if data_is_not_empty:
+            raise ValueError('The data field of hitlets must be empty!')
+
+        data_field_not_long_enough = len(hitlets[0]['data']) < hitlets['length'].max()
+        if data_field_not_long_enough:
+            raise ValueError('The data field must be as large as the longest hitlet in our data.')
+
+        hitlets_with_data_field = hitlets
+    else:
+        n_samples = max(100, hitlets['length'].max())
+        hitlets_with_data_field = np.zeros(len(hitlets), strax.hitlet_with_data_dtype(n_samples))
+        strax.copy_to_buffer(hitlets,
+                             hitlets_with_data_field,
+                             '_copy_hitlets_to_hitlets_width_data')
+
+    _get_hitlets_data(hitlets_with_data_field, records, to_pe, channel_offset)
+    return hitlets_with_data_field
+
+
+@numba.jit(nopython=True, nogil=True, cache=True)
+def _get_hitlets_data(hitlets, records, to_pe, channel_offset):
+    rranges = _touching_windows(records['time'],
+                                strax.endtime(records),
+                                hitlets['time'],
+                                strax.endtime(hitlets))
+
+    for i, h in enumerate(hitlets):
+        recorded_samples_offset = 0
+        n_recorded_samples = 0
+        for ind, r_ind in enumerate(range(rranges[i][0], rranges[i][1])):
+            r = records[r_ind]
+            if r['channel'] != h['channel']:
+                continue
+
+            (r_start, r_end), (h_start, h_end) = strax.overlap_indices(
+                r['time'] // r['dt'],
+                r['length'],
+                h['time'] // h['dt'],
+                h['length'])
+
+            is_first_record = ind == 0
+            if is_first_record:
+                recorded_samples_offset = h_start
+            h_start -= recorded_samples_offset
+            h_end -= recorded_samples_offset
+
+            h['data'][h_start: h_end] += r['data'][r_start: r_end] + r['baseline'] % 1
+            n_recorded_samples += r_end - r_start
```

Comment:
Can you give this another check? This feels wrong to me since the overlapping index is already enforcing the length in hitlet is respected during data fetching.

Reply:
I mean the length field doesn't need to be updated. At least not any case I can think of.

Reply:
You are right, but let me explain as the intention is different. In a few cases we find hits close to the edge of our recorded pulse such that the left or right hit extension reaches beyond the recorded data. As the data is not defined in those region be definition we (nveto subgroup) decided to chop the hitlets in those cases. This only affects less then 1% of our data.